### PR TITLE
Make tests compatible with windows

### PR DESCRIPTION
**Pull request by ocaballeror**
Continuing the discussion from #147 .

Also, just to be clear, I don't intend to merge this yet, so there may be some weird comments and print statements around the code.

### CODE REVIEW

**Comment by ocaballeror** [comment]
Also, it turns out many tests were failing due to timeouts because we were trying to run 8 parallel threads when the appveyor machine has only 2 cores. Using `pytest -n2` works much better :+1:

**Comment by Peque** [comment]
@ocaballeror Yeah, we could even avoid using `pytest-xdist` in AppVeyor. Historically it has given some problems in Travis, but at least it sped things up. If we are not going to notice the speed-up much, then we might as well avoid using it.

I'd say we could even, for AppVeyor, run tests in 1 thread (i.e.: no `pytest-xdist`) and only against the latest Python version.

**Comment by ocaballeror** [comment]
@Peque I've run it a few times with 2 threads and it seems to work fine. We could leave it with 2 and revert to single-threaded mode if we run into any problems in the future. After all, if we have extra cores, why not use them?

About the  python versions, though, why test only the latest one? There could be some implementation differences from version to version that suddenly break compatibility, so I'd leave it as it is right now just to be safe.

**Comment by ocaballeror** [comment]
Update: Still working on the linger tests. There must be something that Windows is doing differently in the internal implementation of sockets and the SO_LINGER option, but I'm having a hard time figuring out what it is.

I'll keep updating when I find more stuff.

**Comment by ocaballeror** [comment]
Took me a long while to figure it out. In the end it turns out the problem with the linger tests had nothing to do with sockets, zmq or anything related to that. The problem was actually being caused by the way we changed the linger value in every test. We were updating the global `config` dictionary, and then expecting the `.close_all()` method to read the updated value, but that won't work on Windows because of the way it creates new processes. On Unix the `fork()` system called is used to create the agent processes, but since that is not available on Windows, and the new processes are created using `spawn()`, we can't expect the global variable `config` to be copied to child processes.

In the end, the call to `get_linger()` inside `close_all()` was returning the default value of 1000 for `linger` instead of the updated one. I changed this so `,close_all()` accepts a `linger` parameter.

**Comment by ocaballeror** [comment]
By the way, there are 2 other tests currently failing and they are both related to Windows allowing more than one socket to listen on one address.

I will try to implement some kind of check to stop this from happening and manually raise an error if the user tries to do that. @Peque or should I just skip these tests on Windows and kindly advise people against this sort of practice?

File: osbrain/tests/test_agent_transport.py [hunk 1]
```diff
@@ -24,18 +25,19 @@ def test_agent_bind_transport_global(nsproxy):
...
 
-    osbrain.config['TRANSPORT'] = 'ipc'
-    agent = run_agent('a2')
-    address = agent.bind('PUSH')
-    assert address.transport == 'ipc'
+    if os.name == 'posix':
+        osbrain.config['TRANSPORT'] = 'ipc'
+        agent = run_agent('a2')
+        address = agent.bind('PUSH')
+        assert address.transport == 'ipc'
```
Comment (Peque) [inline osbrain/tests/test_agent_transport.py]: I think there should probably be a test for global transport configuration, in example:

```python
def test_agent_bind_transport_platform_default(nsproxy):
    """
    Default transport is platform-dependent.
    """
    agent = run_agent('a0')
    address = agent.bind('PUSH')
    if os.name == 'posix':
        assert address.transport == 'ipc'
    else:
        assert address.transport == 'tcp'
```

And then another for the global setting:

```python
def test_agent_bind_transport_global(nsproxy):
    """
    Test global default transport change.
    """
    # Default transport is not `inproc`
    agent = run_agent('a0')
    address = agent.bind('PUSH')
    assert address.transport != 'inproc'

    # Changing default global transport to `inproc`
    osbrain.config['TRANSPORT'] = 'inproc'
    agent = run_agent('a1')
    address = agent.bind('PUSH')
    assert address.transport == 'inproc'
```

File: osbrain/tests/test_agent.py [hunk 2]
```diff
@@ -292,7 +294,7 @@ def on_init(self):
     assert agent_dies('puller', nsproxy)
 
     pusher.send('push', 'foo')
-    pusher.close_all()
+    pusher.close_all(get_linger())
```
Comment (Peque) [inline osbrain/tests/test_agent.py]: This should be removed (see comments on linger changes).

File: osbrain/tests/test_agent_transport.py [hunk 3]
```diff
@@ -46,9 +48,10 @@ def test_agent_bind_transport_agent(nsproxy):
     address = agent.bind('PUSH')
     assert address.transport == 'tcp'
 
-    agent = run_agent('a1', transport='ipc')
-    address = agent.bind('PUSH')
-    assert address.transport == 'ipc'
+    if os.name == 'posix':
+        agent = run_agent('a1', transport='ipc')
+        address = agent.bind('PUSH')
+        assert address.transport == 'ipc'
```
Comment (Peque) [inline osbrain/tests/test_agent_transport.py]: Just use `inproc` instead of `ipc` to make it work in all platforms.

File: osbrain/agent.py [hunk 4]
```diff
@@ -1759,15 +1759,15 @@ def close(self, alias):
...
+    def close_all(self, linger=get_linger()):
         """
         Close all non-internal zmq sockets.
         """
         # Each socket might be pointed by different keys
         sockets_to_delete = []
         for sock in self._get_unique_external_zmq_sockets():
             sockets_to_delete.append(sock)
-            sock.close(linger=get_linger())
+            sock.close(linger)
```
Comment (Peque) [inline osbrain/agent.py]: This is not the expected behavior.

Do you know when that `get_linger()` gets executed?

In my opinion we should leave it as it was before and skip if Windows fails, then create a new issue to investigate this issue and/or a workaround (i.e.: serializing the global configuration to child processes?). But for now it is easier if we simply skip.

File: osbrain/tests/test_proxy.py [hunk 5]
```diff
@@ -57,7 +57,7 @@ def test_wrong_nameserver_address(timeout):
     t0 = time.time()
     with pytest.raises(TimeoutError):
         locate_ns('127.0.0.1:1234', timeout=timeout)
-    assert timeout <= time.time() - t0 <= timeout + 1.
+    assert timeout <= time.time() - t0 <= timeout + 2.
```
Comment (Peque) [inline osbrain/tests/test_proxy.py]: Was this failing with the previous timeout?

File: osbrain/tests/test_proxy.py [hunk 6]
```diff
@@ -323,4 +323,4 @@ def shoot(self):
     assert time.time() - t0 < 0.2
     assert not wayne._next_oneway
 
-    assert wait_agent_attr(wayne, value=20*['bang!'], timeout=1.2)
+    assert wait_agent_attr(wayne, value=20*['bang!'], timeout=1.5)
```
Comment (Peque) [inline osbrain/tests/test_proxy.py]: Was this failing with the previous timeout?

File: osbrain/tests/test_nameserver.py [hunk 7]
```diff
@@ -342,8 +344,10 @@ def test_random_nameserver_process():
     # Raising exceptions
     with pytest.raises(ValueError):
         random_nameserver_process(port_start=-1, port_stop=-2)
-    with pytest.raises(RuntimeError):
-        random_nameserver_process(port_start=22, port_stop=22, timeout=0.5)
+    if sys.platform != 'win32':
+        with pytest.raises(RuntimeError):
+            random_nameserver_process(port_start=22,
+                                      port_stop=22, timeout=0.5)
```
Comment (Peque) [inline osbrain/tests/test_nameserver.py]: I'd say we simply skip this test on Windows for now and revert the changes.

File: osbrain/tests/test_agent_transport.py [hunk 8]
```diff
@@ -71,10 +74,12 @@ def test_agent_bind_given_address(nsproxy):
...
-    ipc_addr = str(uuid4())
-    address = agent.bind('PUSH', addr=ipc_addr, transport='ipc')
-    assert address.transport == 'ipc'
-    assert address.address.name == ipc_addr
+    if os.name == 'posix':
+        ipc_addr = str(uuid4())
+        address = agent.bind('PUSH', addr=ipc_addr, transport='ipc')
+        assert address.transport == 'ipc'
+        assert address.address.name == ipc_addr
+
```
Comment (Peque) [inline osbrain/tests/test_agent_transport.py]: I'd say we split this test in two: one for IPC, one for TCP. Add then a skip-on-windows mark for the IPC test.

**Comment by Peque** [comment]
@ocaballeror I would say, for now, we just skip failing tests for Windows. Then we can create specific issues for those fails and decide whether we should fix them or just assume they should fail in Windows.

Also, that way we keep this PR a bit simpler and merge it faster (even if it is not perfect, it is definitely a step forward that we want to include in our code base). :blush:

File: osbrain/tests/test_agent.py [hunk 9]
```diff
@@ -23,17 +23,19 @@
...
 from common import set_received
+from common import skip_on_windows
 
 
 def test_agent_uuid():
     """
     All agent identifiers should be unique strings.
     """
-    N = 1000
+    N = 100
```
[the remark this answers is not in the document]
Comment (ocaballeror) [reply osbrain/tests/test_agent.py]: Yes. I don't understand why, but for some reason, it takes 12-13 seconds to complete on Windows and about 0.2 seconds on Linux. I thought `N=100` would be enough for this test anyway

File: osbrain/tests/test_agent_transport.py (hunk 1 above)
Comment (ocaballeror) [reply osbrain/tests/test_agent_transport.py]: That reminds me that changing values in the global config does not work on windows, for the same reason I explained earlier when I was talking about the linger thing.
When calling `run_agent()` a new process will be created, but because of the way Windows creates child processes, the values in `config` won't be the ones we just set, but rather the default ones.

I'll add these two tests and try to find a way to make the second one work on Windows.

File: osbrain/tests/test_agent.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Peque) [reply osbrain/tests/test_agent.py]: I knew Windows processes were slow, but did not know they were *that* slow. The main difference is the way they are created (spawn vs. fork). Forking is definitely much faster, specially with the copy-on-write implemented in Linux.

Leave it with 100 then. :+1:

File: osbrain/tests/test_agent_transport.py (hunk 3 above)
Comment (ocaballeror) [reply osbrain/tests/test_agent_transport.py]: Oh. I was convinced 'IPC' and 'inproc' were the same thing. Good thing you pointed it out.

File: osbrain/tests/test_agent_transport.py (hunk 1 above)
Comment (Peque) [reply osbrain/tests/test_agent_transport.py]: As I mention bellow, I'd leave that until we merge this PR. Just having AppVeyor integrated, even if we are skipping some tests in Windows, is worth it to merge. Then we can see how to gradually reduce the number of skipped tests.

File: osbrain/tests/test_proxy.py (hunk 5 above)
Comment (ocaballeror) [reply osbrain/tests/test_proxy.py]: At least on my machine, yes. :smile:

File: osbrain/tests/test_proxy.py (hunk 6 above)
Comment (ocaballeror) [reply osbrain/tests/test_proxy.py]: Maybe at some point, I wouldn't have changed it otherwise. I'll run it through appveyor again, maybe it needed longer timeouts before, since we were using 8 threads instead of 2.

File: osbrain/tests/test_nameserver.py (hunk 7 above)
Comment (ocaballeror) [reply osbrain/tests/test_nameserver.py]: Makes sense.

File: osbrain/tests/test_agent_transport.py (hunk 3 above)
Comment (Peque) [reply osbrain/tests/test_agent_transport.py]: `inproc` is for in-process communication (i.e.: between threads). Happens in memory, so it is much faster than IPC.

File: osbrain/tests/test_nameserver.py (hunk 7 above)
Comment (ocaballeror) [reply osbrain/tests/test_nameserver.py]: What about `random_nameserver_process()` then? Shouldn't we test it works?

File: osbrain/agent.py (hunk 4 above)
Comment (ocaballeror) [reply osbrain/agent.py]: Why is it not the expected behavior? I don't understand what you mean. The value of linger can be passed as an optional parameter, and if it isn't, `get_linger()` is called, as it was before. How does this differ from the expected behavior?

File: osbrain/tests/test_proxy.py (hunk 5 above)
Comment (ocaballeror) [reply osbrain/tests/test_proxy.py]: Turns out 1.2 is enough, but with just 1, the test will fail everytime, both on my machine and appveyor

**Comment by ocaballeror** [comment]
@Peque Maybe a should create another flag? Something like `skip_for_now` so we know that specific test still needs reviewing?

File: osbrain/agent.py [hunk 10]
```diff
@@ -1816,7 +1816,10 @@ def run(self):
 
         try:
             ns = NSProxy(self.nsaddr)
+            reuse = Pyro4.config.SOCK_REUSE
+            Pyro4.config.SOCK_REUSE = False
             self._daemon = Pyro4.Daemon(self._host, self.port)
+            Pyro4.config.SOCK_REUSE = reuse
```
[the remark this answers is not in the document]
Comment (ocaballeror) [reply osbrain/agent.py]: It defaults to `True`. And for some reason (I haven't investigated much), everything fails spectacularly if the default value is changed to `False` in `__init__.py`.

But OK, I'll revert these two changes and skip the tests that rely on OS errors being raised from binding to addresses in use.

File: osbrain/tests/test_nameserver.py (hunk 7 above)
Comment (Peque) [reply osbrain/tests/test_nameserver.py]: For now let us just skip those failing tests. We will fix them later, as separate PRs, analyzing the situation for each case.

**Comment by Peque** [comment]
@ocaballeror I fear the `skip_for_now` name... :joy: (you know, sometimes "temporary" fixes end up being permanent...)

I like the idea of using more specific flags though. Maybe `skip_windows_ipc` and a generic `skip_windows_broken` for (currently) unknown problems.

File: osbrain/agent.py (hunk 4 above)
Comment (Peque) [reply osbrain/agent.py]:
```python
conf = dict(linger=1)


def get_linger():
    return conf['linger']


def default_conf(linger=get_linger()):
    return linger


if __name__ == '__main__':
    print(default_conf())
    conf['linger'] = 2
    print(default_conf())
```

File: osbrain/agent.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ocaballeror) [reply osbrain/agent.py]: Sure

File: osbrain/agent.py (hunk 4 above)
Comment (ocaballeror) [reply osbrain/agent.py]: Oh. I didn't know that happened. What about something like this:
```python
conf = dict(linger=1)


def get_linger():
    return conf['linger']


def default_conf(linger=None):
    if linger is None:
        linger = get_linger()
    return linger


if __name__ == '__main__':
    print(default_conf())
    conf['linger'] = 2
    print(default_conf())
```

Comment (Peque) [reply osbrain/agent.py]: Yeah, although that should be in a separate PR. Basically you'd be adding support for passing a linger parameter to .close_all(). That would require new tests. If you want to implement this we should also probably add the same parameter to .close() (with tests again).

As it is a non-trivial change, I would not add it here. Let us for now keep this PR for only making AppVeyor succeed (even if skipping some tests for now).

Comment (ocaballeror) [reply osbrain/agent.py]: Ok. I will skip the linger tests as well, then.

File: osbrain/tests/common.py [hunk 11]
```diff
@@ -1,10 +1,16 @@
...
 
 from osbrain import run_agent
 from osbrain import run_logger
 from osbrain import run_nameserver
 from osbrain.helper import sync_agent_logger
 
+skip_on_windows = pytest.mark.skipif(sys.platform == 'win32',
+                                     reason='Not supported on windows')
+skip_windows_addr = pytest.mark.skipif(sys.platform == 'win32',
+                                       reason='Windows allows port reuse')
```
Comment (Peque) [inline osbrain/tests/common.py]: Just for consistency I'd name both `skip_on_windows...` or `skip_windows...` (it is easier to `grep`).

File: osbrain/tests/test_agent.py [hunk 12]
```diff
@@ -599,16 +603,18 @@ def test_running_exception(nsproxy):
...
     Running an agent should raise an error if address is already in use.
     """
     with pytest.raises(RuntimeError) as error:
         run_agent('a0', nsaddr=nsproxy.addr(), addr=nsproxy.addr())
     assert 'OSError' in str(error.value)
-    assert 'Address already in use' in str(error.value)
 
 
+@pytest.mark.skipif(sys.platform == 'win32',
+                    reason='Windows allows binding to any port')
```
Comment (Peque) [inline osbrain/tests/test_agent.py]: Even if only used once, I'd declare all Windows skips the same way (with a `@skip...` decorator). Will make finding all Windows skips easier.

File: osbrain/tests/test_agent_transport.py (hunk 1 above)
Comment (Peque) [reply osbrain/tests/test_agent_transport.py]: You forgot to change this (i.e.: create 2 separate tests). :wink:

File: osbrain/tests/test_agent_transport.py (hunk 8 above)
Comment (Peque) [reply osbrain/tests/test_agent_transport.py]: I still think we should split this in 2.

File: osbrain/tests/test_agent_transport.py [hunk 13]
```diff
@@ -90,6 +94,7 @@ def test_agent_bind_given_address(nsproxy):
     assert address.address == tcp_addr
 
 
+@skip_on_windows
```
Comment (Peque) [inline osbrain/tests/test_agent_transport.py]: We could use `skip_on_windows_ipc` to be more clear.

File: osbrain/tests/test_nameserver.py [hunk 14]
```diff
@@ -346,16 +349,18 @@ def test_random_nameserver_process():
...
     Name server start() should raise an error if address is already in use.
     """
     with pytest.raises(RuntimeError) as error:
         run_nameserver(nsproxy.addr())
     assert 'OSError' in str(error.value)
-    assert 'Address already in use' in str(error.value)
 
 
+@pytest.mark.skipif(sys.platform == 'win32',
+                    reason='Windows allows binding to any port')
```
Comment (Peque) [inline osbrain/tests/test_nameserver.py]: Rather use `@skip...` everywhere.

File: osbrain/tests/test_agent_ipc_sockets.py [hunk 15]
```diff
@@ -9,6 +10,9 @@
 
 from common import nsproxy  # pragma: no flakes
 
+pytestmark = pytest.mark.skipif(sys.platform == 'win32',
+                                reason='IPC not available on Windows')
```
Comment (Peque) [inline osbrain/tests/test_agent_ipc_sockets.py]: Does `pytestmark = skip_windows_ipc` work? (importing the corresponding `skip_windows_ipc` definition).

This way we put all `pytest.mark.skipif` declarations together.

File: osbrain/tests/test_nameserver.py [hunk 16]
```diff
@@ -346,16 +349,17 @@ def test_random_nameserver_process():
...
 
+@skip_windows_port_reuse
 def test_nameserver_oserror(nsproxy):
     """
     Name server start() should raise an error if address is already in use.
     """
     with pytest.raises(RuntimeError) as error:
         run_nameserver(nsproxy.addr())
     assert 'OSError' in str(error.value)
-    assert 'Address already in use' in str(error.value)
```
Comment (Peque) [inline osbrain/tests/test_nameserver.py]: You removed this line but now that we are skipping for Windows, it should be there.

File: osbrain/tests/test_agent.py [hunk 17]
```diff
@@ -301,7 +306,7 @@ def on_init(self):
     puller.bind('PULL', alias='pull', handler=append_received,
                 addr=address.address, transport='tcp')
 
-    assert should_receive == wait_agent_attr(puller, data='foo', timeout=.2)
+    assert should_receive == wait_agent_attr(puller, data='foo', timeout=1)
```
Comment (Peque) [inline osbrain/tests/test_agent.py]: I guess this timeout does not need to be changed for now as this test is skipped for Windows. When/if we remove the skip we will change the timeout if necessary.

File: osbrain/tests/test_agent.py [hunk 18]
```diff
@@ -599,16 +604,17 @@ def test_running_exception(nsproxy):
...
 
+@skip_windows_port_reuse
 def test_agent_error_address_already_in_use(nsproxy):
     """
     Running an agent should raise an error if address is already in use.
     """
     with pytest.raises(RuntimeError) as error:
         run_agent('a0', nsaddr=nsproxy.addr(), addr=nsproxy.addr())
     assert 'OSError' in str(error.value)
-    assert 'Address already in use' in str(error.value)
```
Comment (Peque) [inline osbrain/tests/test_agent.py]: You removed this line but now that we are skipping for Windows, it should be there.

File: osbrain/tests/test_proxy.py (hunk 6 above)
Comment (Peque) [reply osbrain/tests/test_proxy.py]: Is still fails if set back to `1.2`?

File: osbrain/tests/test_agent.py (hunk 17 above)
Comment (ocaballeror) [reply osbrain/tests/test_agent.py]: But this one does fail on Linux from time to time. I ran it in a loop yesterday, and it took me anywhere between 50 and 500 attempts to make it fail, but eventually it does.

It has failed a few times in travis:
https://travis-ci.org/ocaballeror/osbrain/jobs/346269407

I think 1 is a safer value.

Comment (Peque) [reply osbrain/tests/test_agent.py]: @ocaballeror Then ok. It seems changing the timeout does not actually change the test, will only make it a bit slower for the cases where `should_receive == False`, so it is fine for me. :blush:

Should we put it as a separate commit? i.e.: "More lenient timeout in linger test". Simply because it is not really related to Windows tests.

Comment (ocaballeror) [reply osbrain/tests/test_agent.py]: Makes sense 👍 

Rebasing this is going to be fun....

File: osbrain/tests/test_proxy.py (hunk 6 above)
Comment (ocaballeror) [reply osbrain/tests/test_proxy.py]: Same as above, I ran it in a loop to see if it would eventually fail, and after 100-200 tests I've gotten it to fail a couple of times. 1.5 looks like a safer value.

**Comment by ocaballeror** [comment]
Finally 😄 

This took way longer than I expected
